File: frontend/src/types/model_provider.ts
// Provider
export interface ModelProviderResponse {
  id: string
  name: string
  base_url: string | null
  has_api_key: boolean // Never expose actual key
  is_active: boolean
  models: ModelProviderModelResponse[]
  created_at: string
  updated_at: string | null
}

export interface ModelProviderCreate {
  name: string
  base_url?: string
  api_key?: string
  is_active?: boolean
}

export type ModelProviderUpdate = Partial<ModelProviderCreate>

// Provider Model (child of Provider)
export interface ModelProviderModelResponse {
  id: string
  provider_id?: string // May be omitted when nested under a provider response
  name: string // Display name: "DeepSeek V3"
  model_name: string // API identifier: "deepseek-chat"
  temperature: number | null
  max_output_tokens: number | null
  context_size: number | null
  json_mode_enabled: boolean
  tool_choice_enabled: boolean
  supports_vision: boolean
  is_active: boolean
  created_at: string
  updated_at: string | null
}

export interface ModelProviderModelCreate {
  name: string
  model_name: string
  temperature?: number
  max_output_tokens?: number
  context_size?: number
  json_mode_enabled?: boolean
  tool_choice_enabled?: boolean
  supports_vision?: boolean
}

export type ModelProviderModelUpdate = Partial<
  ModelProviderModelCreate & { is_active: boolean; supports_vision: boolean }
>

// Model Group (Profile)
export interface ModelGroupResponse {
  id: string
  name: string
  description: string | null
  general_model_id: string | null
  fast_model_id: string | null
  reasoning_model_id: string | null
  general_model: ModelSlotInfo | null
  fast_model: ModelSlotInfo | null
  reasoning_model: ModelSlotInfo | null
  is_active: boolean
  created_at: string
  updated_at: string | null
}

export interface ModelSlotInfo {
  id: string
  name: string
  model_name: string
  provider_name: string
  is_available: boolean
}

export interface ModelGroupCreate {
  name: string
  description?: string
  general_model_id?: string
  fast_model_id?: string
  reasoning_model_id?: string
}

export type ModelGroupUpdate = Partial<ModelGroupCreate>

// Active Configuration
export interface ModelActiveConfig {
  mode: "env" | "group"
  active_group: { id: string; name: string } | null
  effective: {
    general: EffectiveModel
    fast: EffectiveModel
    reasoning: EffectiveModel
  }
  env_fallback: EnvFallbackInfo
}

export interface EffectiveModel {
  model_name: string
  provider_name: string | null
  source: "group" | "env"
}

export interface EnvFallbackInfo {
  llm_model: string
  llm_base_url: string
  fast_llm_model: string
  fast_llm_base_url: string
  reasoning_llm_model: string
  reasoning_llm_base_url: string
  has_api_key: boolean
  has_fast_api_key: boolean
  has_reasoning_api_key: boolean
}

// List responses
export interface ModelProvidersListResponse {
  providers: ModelProviderResponse[]
  total: number
}

export interface ModelGroupsListResponse {
  groups: ModelGroupResponse[]
  env_fallback: EnvFallbackInfo
  active_group_id: string | null
}

File: src/fim_one/web/schemas/model_provider.py
"""Pydantic schemas for model provider / model / group endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Provider schemas
# ---------------------------------------------------------------------------


class ProviderCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    base_url: str = Field(min_length=1, max_length=500)
    api_key: str | None = None
    is_active: bool = True


class ProviderUpdate(BaseModel):
    name: str | None = None
    base_url: str | None = None
    api_key: str | None = None
    is_active: bool | None = None


class ProviderModelResponse(BaseModel):
    """A model entry as nested inside a provider response."""

    id: str
    name: str
    model_name: str
    temperature: float | None
    max_output_tokens: int | None
    context_size: int | None
    json_mode_enabled: bool
    tool_choice_enabled: bool
    supports_vision: bool
    is_active: bool
    created_at: str
    updated_at: str | None


class ProviderResponse(BaseModel):
    id: str
    name: str
    base_url: str | None
    has_api_key: bool
    is_active: bool
    models: list[ProviderModelResponse]
    created_at: str
    updated_at: str | None
    # NEVER expose api_key in responses


class ProviderListResponse(BaseModel):
    providers: list[ProviderResponse]
    total: int


# ---------------------------------------------------------------------------
# Provider Model schemas
# ---------------------------------------------------------------------------


class ProviderModelCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    model_name: str = Field(min_length=1, max_length=100)
    temperature: float | None = None
    max_output_tokens: int | None = None
    context_size: int | None = None
    json_mode_enabled: bool = True
    tool_choice_enabled: bool = True
    supports_vision: bool = False


class ProviderModelUpdate(BaseModel):
    name: str | None = None
    model_name: str | None = None
    temperature: float | None = None
    max_output_tokens: int | None = None
    context_size: int | None = None
    json_mode_enabled: bool | None = None
    tool_choice_enabled: bool | None = None
    supports_vision: bool | None = None
    is_active: bool | None = None


class ProviderModelFullResponse(BaseModel):
    """A model entry with its provider info included."""

    id: str
    provider_id: str
    provider_name: str
    name: str
    model_name: str
    temperature: float | None
    max_output_tokens: int | None
    context_size: int | None
    json_mode_enabled: bool
    tool_choice_enabled: bool
    supports_vision: bool
    is_active: bool
    created_at: str
    updated_at: str | None


# ---------------------------------------------------------------------------
# Model Group schemas
# ---------------------------------------------------------------------------


class ModelSlotInfo(BaseModel):
    """Resolved info for a model slot (general/fast/reasoning).

    Field names match the frontend ``ModelSlotInfo`` TypeScript interface.
    """

    id: str
    name: str  # Display name of the model
    model_name: str  # API model identifier
    provider_name: str  # Provider display name
    is_available: bool  # True if both model and its provider are active


class GroupCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = None
    general_model_id: str | None = None
    fast_model_id: str | None = None
    reasoning_model_id: str | None = None


class GroupUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    general_model_id: str | None = None
    fast_model_id: str | None = None
    reasoning_model_id: str | None = None


class GroupResponse(BaseModel):
    id: str
    name: str
    description: str | None
    general_model_id: str | None
    fast_model_id: str | None
    reasoning_model_id: str | None
    general_model: ModelSlotInfo | None
    fast_model: ModelSlotInfo | None
    reasoning_model: ModelSlotInfo | None
    is_active: bool
    created_at: str
    updated_at: str | None


class EnvFallbackInfoV2(BaseModel):
    """Extended ENV fallback info including reasoning tier."""

    llm_model: str
    llm_base_url: str
    fast_llm_model: str
    fast_llm_base_url: str
    reasoning_llm_model: str
    reasoning_llm_base_url: str
    has_api_key: bool
    has_fast_api_key: bool
    has_reasoning_api_key: bool


class GroupListResponse(BaseModel):
    groups: list[GroupResponse]
    env_fallback: EnvFallbackInfoV2
    active_group_id: str | None


# ---------------------------------------------------------------------------
# Active Configuration schema
# ---------------------------------------------------------------------------


class EffectiveModelInfo(BaseModel):
    model_name: str | None
    provider_name: str | None
    source: str  # "group" or "env"


class ActiveConfigResponse(BaseModel):
    mode: str  # "env" or "group"
    active_group: dict[str, Any] | None  # { id, name } or null
    effective: dict[str, EffectiveModelInfo]  # keys: general, fast, reasoning
    env_fallback: EnvFallbackInfoV2


# ---------------------------------------------------------------------------
# Model Config Import / Export schemas
# ---------------------------------------------------------------------------


class ModelExportData(BaseModel):
    """A single model entry in the export payload."""

    name: str
    model_name: str
    temperature: float | None = None
    max_output_tokens: int | None = None
    context_size: int | None = None
    json_mode_enabled: bool = True
    tool_choice_enabled: bool = True
    supports_vision: bool = False
    is_active: bool = True


class ProviderExportData(BaseModel):
    """A provider with its models in the export payload."""

    name: str
    base_url: str | None = None
    api_key: None = None  # NEVER export secrets
    is_active: bool = True
    models: list[ModelExportData] = []


class GroupModelRef(BaseModel):
    """Portable reference to a model by provider name + model_name."""

    provider: str
    model_name: str


class GroupExportData(BaseModel):
    """A model group in the export payload."""

    name: str
    description: str | None = None
    general_model: GroupModelRef | None = None
    fast_model: GroupModelRef | None = None
    reasoning_model: GroupModelRef | None = None
    is_active: bool = False


class ModelConfigExportEnvelope(BaseModel):
    """The inner envelope for model config export."""

    exported_at: str
    providers: list[ProviderExportData]
    groups: list[GroupExportData]


class ModelConfigExportResponse(BaseModel):
    """Top-level export wrapper with versioned key."""

    fim_model_config_v1: ModelConfigExportEnvelope


class ModelConfigImportRequest(BaseModel):
    """Request body for model config import."""

    fim_model_config_v1: ModelConfigExportEnvelope
    api_keys: dict[str, str] = Field(
        default_factory=dict,
        description="Optional mapping of provider name -> API key",
    )
    clear_existing: bool = Field(
        default=False,
        description="If true, delete ALL existing providers, models, and groups before importing",
    )


class ModelConfigImportSummary(BaseModel):
    """Summary of import results."""

    created: dict[str, int]  # { providers, models, groups }
    skipped: dict[str, int]  # { providers, models, groups }
    deleted: dict[str, int] = Field(
        default_factory=lambda: {"providers": 0, "models": 0, "groups": 0}
    )
    warnings: list[str] = []
